Index: loader/Makefile
===================================================================
--- loader/Makefile	(revision 23850)
+++ loader/Makefile	(working copy)
@@ -29,6 +29,7 @@
                dmo/dmo.c  \
                dmo/dmo_guids.c \
 
+LIBS-yes += registercodec
 include ../mpcommon.mak
 
 CFLAGS+=-fno-omit-frame-pointer -mno-omit-leaf-frame-pointer
@@ -36,9 +37,13 @@
 dshow/test: dshow/test.c $(LIBNAME)
 	$(CC) $(CFLAGS) -o $@ $^ -lstdc++
 
+registercodec: modify_reg.c registry.c
+	$(CC) $(CFLAGS) -o $@ $<
+
 clean::
 	rm -f dshow/*.o dshow/*.a dshow/*~
 	rm -f dmo/*.o dmo/*.a dmo/*~
 
 distclean::
 	rm -f dshow/test dshow/test.raw
+	rm -f registercodec
